common/src/log.rs

```diff
@@ -1,123 +1,172 @@
-        };
-
         write!(
             f,
-            "{} {} {}",
+            "{} [{}] {}",
             datetime.to_rfc3339().dim(),
-            self.level.get_colored(),
-            message
+            self.internal_origin,
+            self.line,
         )
     }
 }
 
-#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
-#[serde(rename_all = "lowercase")]
-#[cfg_attr(feature = "openapi", derive(ToSchema))]
-#[cfg_attr(feature = "openapi", schema(as = shuttle_common::log::Level))]
-pub enum Level {
-    Trace,
-    Debug,
-    Info,
-    Warn,
-    Error,
+/// Records logs for the deployment progress
+pub trait LogRecorder: Clone + Send + 'static {
+    fn record(&self, log: LogItem);
 }
 
-#[cfg(feature = "display")]
-impl Level {
-    fn get_colored(&self) -> StyledContent<&str> {
-        match self {
-            Level::Trace => "TRACE".magenta(),
-            Level::Debug => "DEBUG".blue(),
-            Level::Info => " INFO".green(),
-            Level::Warn => " WARN".yellow(),
-            Level::Error => "ERROR".red(),
+/// Tracing subscriber layer which logs based on if the log
+/// is from a span that is tagged with a deployment id
+pub struct DeploymentLogLayer<R>
+where
+    R: LogRecorder + Send + Sync,
+{
+    pub recorder: R,
+    pub internal_service: Backend,
+}
+
+impl<R, S> Layer<S> for DeploymentLogLayer<R>
+where
+    S: Subscriber + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
+    R: LogRecorder + Send + Sync + 'static,
+{
+    fn on_event(&self, event: &tracing::Event<'_>, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        // We only care about events in some scope
+        let scope = if let Some(scope) = ctx.event_scope(event) {
+            scope
+        } else {
+            return;
+        };
+
+        // Find the outermost scope with the scope details containing the current state
+        for span in scope.from_root() {
+            let extensions = span.extensions();
+
+            if let Some(details) = extensions.get::<ScopeDetails>() {
+                self.recorder.record(LogItem {
+                    id: details.id,
+                    internal_origin: self.internal_service.clone(),
+                    timestamp: Utc::now(),
+                    line: "Test".into(),
+                });
+                break;
+            }
         }
     }
+    fn on_new_span(
+        &self,
+        attrs: &span::Attributes<'_>,
+        id: &span::Id,
+        ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) {
+        // We only care about spans that change the state
+        if !DeploymentIdVisitor::is_valid(attrs.metadata()) {
+            return;
+        }
+        let mut visitor = DeploymentIdVisitor::default();
+        attrs.record(&mut visitor);
+        let details = visitor.details;
+
+        if details.id.is_nil() {
+            warn!("scope details does not have a valid id");
+            return;
+        }
```

Comment:
Is this even possible? Because of the `is_valid()` check

Reply:
Copied previous implementation. Can check.

Reply:
It is not necessary, but a good sanity check.